src/coreComponents/common/fixedSizeDeque.hpp

```diff
@@ -0,0 +1,158 @@
+/*
+ * ------------------------------------------------------------------------------------------------------------
+ * SPDX-License-Identifier: LGPL-2.1-only
+ *
+ * Copyright (c) 2018-2020 Lawrence Livermore National Security LLC
+ * Copyright (c) 2018-2020 The Board of Trustees of the Leland Stanford Junior University
+ * Copyright (c) 2018-2020 TotalEnergies
+ * Copyright (c) 2019-     GEOSX Contributors
+ * All rights reserved
+ *
+ * See top level LICENSE, COPYRIGHT, CONTRIBUTORS, NOTICE, and ACKNOWLEDGEMENTS files for details.
+ * ------------------------------------------------------------------------------------------------------------
+ */
+#ifndef FIXEDSIZEDEQUE_HPP
+#define FIXEDSIZEDEQUE_HPP
+
+#if defined( GEOSX_USE_CUDA )
+#  include <cuda.h>
+#  include <cuda_runtime.h>
+#endif
+
+#include <chai/ArrayManager.hpp>
+#include <chai/ChaiMacros.hpp>
+#include <string>
+#include <functional>
+#include "LvArray/src/Array.hpp"
+#include "LvArray/src/ArrayView.hpp"
+#include "LvArray/src/memcpy.hpp"
+#include "common/GEOS_RAJA_Interface.hpp"
+#include "common/Logger.hpp"
+
+/// Get the positive value of a module b
+#define POSITIVE_MODULO( a, b ) ( ( ( a ) % ( b ) ) + b ) % ( b )
+namespace geosx
+{
+template< typename T, typename INDEX_TYPE >
+/// Implement a double ended queue with fixed number of fixed size buffer to be stored
+class fixedSizeDeque
+{
+  /// The integer type used for indexing.
+  using IndexType = INDEX_TYPE;
+  /// 1D array slice
+  using ArraySlice1D = LvArray::ArraySlice< T const, 1, 0, INDEX_TYPE >;
```

Comment:
If the home for this stuff is going to be in geosx then I'd remove the `INDEX_TYPE` template parameter and use `localIndex` along with the Array definitions here

https://github.com/GEOSX/GEOSX/blob/15de98b7259aa676d9d8cbcaacfcd30baa6b3347/src/coreComponents/common/DataTypes.hpp#L210

Reply:
Do you think we should move that to LvArray or not ?
By the way, I may add ArrayOfArray based queue if I happen to add compression of the buffer before storage, later on.